### PR TITLE
Support tuples as types

### LINKED ISSUE #6635: Support  multi column IN lists like `(c1, c2) IN ((c1, c2), ,,,)`
### Is your feature request related to a problem or challenge?

A user on slack was using datafusion to query parquet files from S3: https://the-asf.slack.com/archives/C01QUFS30TD/p1686147411917959

They reported that the following predicate got 1000x slower when it had 100,000 distinct filter values:
```sql
WHERE
(rate.code = 'SC1' and rate.type = 'UT1' and rate.operation = 'RO1') 
OR 
(rate.code = 'SC2' and rate.type = 'UT2' and rate.operation = 'RO2')
...
```

However, when it was rewritten as an inlist it went much faster:
```sql
WHERE (rate.code || '_' || rate.type || '_' || rate.operation) IN ('SC1_UT1_RO1', 'SC2_UT2_RO2', ...)
```

However, this rewrite is not general (for example, if `code`, `type` or `operation` contain `_` characters.

SQL supports this type of predicate natively with "multi-column inlists" that look like;

```sql
WHERE (rate.code,  rate.type, rate.operation) IN (('SC1', 'UT1', 'RO1'), ('SC2', 'UT2', 'RO2') ...)
```

 Substrait supports this kind of predicate too, which I take as some evidence it is widely used

https://substrait.io/expressions/specialized_record_expressions/#or-list-equality-expression

```
> A specialized structure that is often used is a large list of possible values. In SQL, these are typically large IN lists. They can be composed from one or more fields. There are two common patterns, single value and multi value. In pseudocode they are represented as:
```

### Describe the solution you'd like

I would like multi-column predicates to work in DataFusion. 

today, they result in an "unimplemented" error:

```sql
❯ create table foo (x int, y varchar);
0 rows in set. Query took 0.001 seconds.
❯ insert into foo values (1, 'a'), (2, 'b'), (3, 'c');
+-------+
| count |
+-------+
| 3     |
+-------+
1 row in set. Query took 0.002 seconds.
❯ select * from foo where (x, y) IN ((1,'a'), (2, 'b'), (5, 'e'));
This feature is not implemented: Unsupported ast node in sqltorel: Tuple([Value(Number("1", false)), Value(SingleQuotedString("a"))])
```

Here is an example showing this feature working in posgres:

```sql
postgres=# create table foo (x int, y varchar);
CREATE TABLE
postgres=# insert into foo values (1, 'a'), (2, 'b'), (3, 'c');
INSERT 0 3
postgres=# select * from foo;
 x | y
---+---
 1 | a
 2 | b
 3 | c
(3 rows)
postgres=# select * from foo where (x, y) IN ((1,'a'), (2, 'b'), (5, 'e'));
 x | y
---+---
 1 | a
 2 | b
(2 rows)
```

### Describe alternatives you've considered

The existing InList structure looks like this: https://docs.rs/datafusion-expr/26.0.0/datafusion_expr/expr/struct.InList.html

```rust
pub struct InList {
    pub expr: Box<Expr>,
    pub list: Vec<Expr>,
    pub negated: bool,
}
```

I am not sure how best to implement this. One idea is to simply special case multi-inputs, something like 

```rust
struct MultIInList {
    pub exprs: Vec<Expr>,
    pub lists: Vec<Vec<Expr>>,
    pub negated: bool,
}
```

However, my preferred approach would be to support `StructArray`s in `InList` and then implement a rewrite from

```sql
(col1, col2, ...) in (('a1', 'a2', ..), ('b1', 'b2', ..), ...)
```
into

```
struct(col1, col2) In ({ col1: 'a', col2: 'a2', ..}, {col1: 'b1', col2: 'b2', ..}, ...)
```

While likely more complicated this approach would then support structs in INLISTs directly which I think will be more and more valuable over time

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by samuelcolvin**
## Which issue does this PR close?

Closes #6635, continues from #8590.

## Rationale for this change

See #6635 — I want to be able to use tuples as per that issue

## What changes are included in this PR?

* Add tuples planning
* add tuple coercion
* tests

## Are these changes tested?

yes.

## Are there any user-facing changes?

More SQL supported.

**(Fixed, ignore this)**

~~This is mostly working and `(a, b) in (('x', 'y'))` is working correctly, however if you have more than one value in the lookup, e.g. `('x', 'y') IN (('x', 'y'), ('y', 'x'))` an error occurs:~~

```
External error: query failed: DataFusion error: Arrow error: Invalid argument error: Nested comparison: Struct([Field { name: "c0", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c1", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }]) == Struct([Field { name: "c0", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c1", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }]) (hint: use make_comparator instead)
[SQL] select ('x', 'y') IN (('x', 'y'), ('y', 'x'));
at test_files/struct.slt:256
```

~~There are two tests that should fail.~~

### CODE REVIEW

File: datafusion/sqllogictest/test_files/struct.slt [hunk 1]
```diff
@@ -236,3 +233,44 @@ query ?
...
+query error DataFusion error: Error during planning: Cannot infer common argument type for comparison operation Struct.*
+select ('x', 'y') = ('x', 'y', 'z');
+
+query B
+select ('x', 'y') IN (('x', 'y'));
+----
+true
+
+query B
+select ('x', 'y') IN (('x', 'y'), ('y', 'x'));
```
Comment (samuelcolvin) [inline datafusion/sqllogictest/test_files/struct.slt:257]: this will fail

File: datafusion/sqllogictest/test_files/struct.slt [hunk 2]
```diff
@@ -236,3 +233,44 @@ query ?
...
+----
+true
+
+query I
+select a from values where (a, c) = (1, 'a');
+----
+1
+
+query I
+select a from values where (a, c) IN ((1, 'a'), (2, 'b'));
```
Comment (samuelcolvin) [inline datafusion/sqllogictest/test_files/struct.slt:267]: this will fail

**Review by samuelcolvin — COMMENTED** [review]
I don't know why

**Comment by jayzhan211** [comment]
`DataFusion error: Arrow error: Invalid argument error: Nested comparison: Struct\(\[Field \{ name: "c0", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: \{\} \}, Field \{ name: "c1", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: \{\} \}\]\) == Struct\(\[Field \{ name: "c0", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: \{\} \}, Field \{ name: "c1", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: \{\} \}\]\) \(hint: use make_comparator instead\)`

`make_comparator ` is introduced recently for comparing nested type.

However, InListExpr had not yet supported to compare nested type
```rust
fn evaluate(&self, batch: &RecordBatch) -> Result<ColumnarValue> {
        let num_rows = batch.num_rows();
        let value = self.expr.evaluate(batch)?;
        let r = match &self.static_filter {
            Some(f) => f.contains(value.into_array(num_rows)?.as_ref(), self.negated)?,
            None => {
                let value = value.into_array(num_rows)?;
                let found = self.list.iter().map(|expr| expr.evaluate(batch)).try_fold(
                    BooleanArray::new(BooleanBuffer::new_unset(num_rows), None),
                    |result, expr| -> Result<BooleanArray> {
                        Ok(or_kleene(
                            &result,
                            &eq(&value, &expr?.into_array(num_rows)?)?,
                        )?)
                    },
                )?;

                if self.negated {
                    not(&found)?
                } else {
                    found
                }
            }
        };
        Ok(ColumnarValue::Array(Arc::new(r)))
    }
```

We need to change the `eq` to new function that calls `compare_op_for_nested` for nested type, and fallback to `eq` for non-nested type.

**Comment by samuelcolvin** [comment]
Thanks so much for the pointer @jayzhan211, this should now be fixed.

File: datafusion/sql/src/expr/mod.rs [hunk 3]
```diff
@@ -695,6 +696,23 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+        schema: &DFSchema,
+        planner_context: &mut PlannerContext,
+        values: Vec<SQLExpr>,
+    ) -> Result<Expr> {
+        match values.first() {
+            Some(SQLExpr::Identifier(_)) | Some(SQLExpr::Value(_)) => {
+                self.parse_struct(schema, planner_context, values, vec![])
+            }
+            None => not_impl_err!("Empty tuple not supported yet"),
+            _ => {
```
Comment (alamb) [inline datafusion/sql/src/expr/mod.rs:710]: I was trying to figure out how to write a negative test for this case, however anything I could try seems to work

DataFusion CLI v41.0.0
> create table foo as values (1);

```sql
> select * from foo where (column1) IN ((column1+1), (2));
+---------+
| column1 |
+---------+
+---------+
0 row(s) fetched.
Elapsed 0.008 seconds.

> select * from foo where (column1) IN ((column1+1), (2), (1));
+---------+
| column1 |
+---------+
| 1       |
+---------+
1 row(s) fetched.
Elapsed 0.010 seconds.
```

File: datafusion/sqllogictest/test_files/struct.slt [hunk 4]
```diff
@@ -236,3 +233,44 @@ query ?
...
+----
+true
+
+query B
+select ('x', 'y') IN (('x', 'y'), ('y', 'x'));
+----
+true
+
+query I
+select a from values where (a, c) = (1, 'a');
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/struct.slt:262]: I recommend we also do a negative test, like 

```
select a from values where (a, c) = (100, 'a');
```

**Review by alamb — APPROVED** [review]
Thank you @samuelcolvin  -- this looks very impressive to me. Beautifully simple

I had some small suggestions, but I don't think they are required to merge

Very nicely done

**Comment by alamb** [comment]
FYI @my-vegetable-has-exploded

File: datafusion/expr/src/type_coercion/binary.rs [hunk 5]
```diff
@@ -780,6 +781,31 @@ fn coerce_numeric_type_to_decimal256(numeric_type: &DataType) -> Option<DataType
...
+
+            let types = std::iter::zip(lhs_fields.iter(), rhs_fields.iter())
+                .map(|(lhs, rhs)| comparison_coercion(lhs.data_type(), rhs.data_type()))
+                .collect::<Option<Vec<DataType>>>()?;
+
+            let fields = types
+                .into_iter()
+                .enumerate()
+                .map(|(i, datatype)| {
+                    Arc::new(Field::new(format!("c{i}"), datatype, true))
```
Comment (jayzhan211) [inline datafusion/expr/src/type_coercion/binary.rs:800]: I think we should preserve field name and nullable, only type is updated

File: datafusion/sql/src/expr/mod.rs (hunk 3 above)
Comment (jayzhan211) [reply datafusion/sql/src/expr/mod.rs:710]:
```rust
query error DataFusion error: This feature is not implemented: Only identifiers and literals are supported in tuples
select a from values where (a, c) IN ((abs(a), 1));
```
Query success in duckdb
```
D select a from t where (a, b) in ((1, abs(a)));
┌────────┐
│   a    │
│ int32  │
├────────┤
│ 0 rows │
└────────┘
```

We could support `abs()` and other complex type in tuple in follow PR, but we should check not only `first` element of the values but all the values in this PR

Comment (alamb) [reply datafusion/sql/src/expr/mod.rs:710]: I will file a ticket to track the feature

Comment (my-vegetable-has-exploded) [reply datafusion/sql/src/expr/mod.rs:710]: > ```rust
> query error DataFusion error: This feature is not implemented: Only identifiers and literals are supported in tuples
> select a from values where (a, c) IN ((abs(a), 1));
> ```
> 
> Query success in duckdb
> 
> ```
> D select a from t where (a, b) in ((1, abs(a)));
> ┌────────┐
> │   a    │
> │ int32  │
> ├────────┤
> │ 0 rows │
> └────────┘
> ```
> 
> We could support `abs()` and other complex type in tuple in follow PR, but we should check not only `first` element of the values but all the values in this PR

I am not sure about whether our evaulation support struct array like `{(struct(Expr1{column1}, 1)), struct (Expr2{column2}, 2))}`?

**Comment by my-vegetable-has-exploded** [comment]
Thanks @samuelcolvin. This is exactly what I wanted to do.

**Comment by alamb** [comment]
I took the liberty of merging up from main to resolve a conflict -- I think we could merge this PR and address the comments as a follow on PR, or we can do it prior to merging this one.

Please let me know what you prefer @samuelcolvin

**Comment by alamb** [comment]
Thanks again everyone 🚀

**Pull request merged**